### PR TITLE
Compute diffID for mapped-layer at creating image source

### LINKED ISSUE #2800: Fail to commit a container with UID and GID mappings
**Description**

A container cannot be committed if UID and GID mapping is used.
`commit` fails because a mapped-layer doesn't have its diffID necessary for looking up the layer size.

**Steps to reproduce the issue:**
1. Pull an image: `buildah pull docker.io/library/alpine`
2. Create a container with UID and GID mapping:
   `buildah from --userns-uid-map 0:100000:65536 --userns-gid-map 0:100000:65536 docker.io/library/alpine`
3. Commit the container: `buildah commit alpine-working-container localhost/myalpine`

With podman:
1. Pull an image: `podman pull docker.io/library/alpine`
2. Create a container with UID and GID mapping:
  `podman run --uidmap 0:100000:65536 --gidmap 0:100000:65536 --name alpine-working-container docker.io/library/alpine`
3. Commit the container: `podman commit alpine-working-container localhost/myalpine`

**Describe the results you received:**

Failed to commit with the error: unable to look up size of layer

```
# buildah pull docker.io/library/alpine
Getting image source signatures
Copying blob 188c0c94c7c5 done
Copying config d6e46aa247 done
Writing manifest to image destination
Storing signatures
d6e46aa2470df1d32034c6707c8041158b652f38d2a9ae3d7ad7e7532d22ebe0
# buildah from --userns-uid-map 0:100000:65536 --userns-gid-map 0:100000:65536 docker.io/library/alpine
alpine-working-container
# buildah commit alpine-working-container localhost/myalpine
error committing container "alpine-working-container" to "localhost/myalpine": error copying layers and metadata for container "d294d7551aca840304e99194c523c64ed153b7084340a8268a23d4b2316e8c0b": Error initializing source containers-storage:alpine-working-container: unable to look up size of layer "1363197c109ea8af28225a442f8b1ea2fd30f7352e3fcf1249f9d5d5ca1c98ea"
```

**Describe the results you expected:**

Succeed in committing.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.15.1-1.fc32.x86_64
```

**Output of `buildah version`:**

```
Version:         1.15.1
Go Version:      go1.14.6
Image Spec:      1.0.1-dev
Runtime Spec:    1.0.2-dev
CNI Spec:        0.4.0
libcni Version:
image Version:   5.5.1
Git Commit:
Built:           Thu Jan  1 09:00:00 1970
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
Version:      2.1.1
API Version:  2.0.0
Go Version:   go1.14.9
Built:        Thu Oct  1 04:31:11 2020
OS/Arch:      linux/amd64
```

**Output of `cat /etc/*release`:**

```
Fedora release 32 (Thirty Two)
NAME=Fedora
VERSION="32 (Server Edition)"
ID=fedora
VERSION_ID=32
VERSION_CODENAME=""
PLATFORM_ID="platform:f32"
PRETTY_NAME="Fedora 32 (Server Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:32"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f32/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=32
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=32
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Server Edition"
VARIANT_ID=server
Fedora release 32 (Thirty Two)
Fedora release 32 (Thirty Two)
```

**Output of `uname -a`:**

```
Linux localhost.localdomain 5.6.6-300.fc32.x86_64 #1 SMP Tue Apr 21 13:44:19 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
[storage.options.overlay]
# ignore_chown_errors can be set to allow a non privileged user running with
# a single UID within a user namespace to run containers. The user can pull
# and use any image even those with multiple uids.  Note multiple UIDs will be
# squashed down to the default uid in the container.  These images will have no
# separation between the users in the container. Only supported for the overlay
# and vfs drivers.
#ignore_chown_errors = "false"

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev,metacopy=on"

# Set to skip a PRIVATE bind mount on the storage home directory.
# skip_mount_home = "false"

# Size is used to set a maximum size of the container image.
# size = ""

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the bas
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base
# device.
# mkfsarg = ""

# metadata_size is used to set the `pvcreate --metadatasize` options when
# creating thin devices. Default is 128k
# metadata_size = ""

# Size is used to set a maximum size of the container image.
# size = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"
```

### PR DESCRIPTION
**Pull request by hshiina**
Signed-off-by: Hironori Shiina <shiina.hironori@fujitsu.com>

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
If UID and GID mappings are specified, the container has a
mapped-layer, whose diffID is not computed when created.
Committing the image fails due to lack of diffID. This fix
computes diffID at creating an image source if a layer
doesn't have a diffID (UncompressedDigest).

This fix also tests if a container with UID and GID mappings
can be committed.

#### How to verify it

This PR adds a test to verify if a container with UID and GID
mappings can be committed to `idmapping` in `namespaces.bats`.

#### Which issue(s) this PR fixes:

Fixes #2800 

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by rhatdan** [comment]
/approve
LGTM
@nalind @TomSweeneyRedHat @vrothberg PTAL

**Review by vrothberg — COMMENTED** [review]
LGTM, thanks!

Want a final head nod from @nalind and/or @TomSweeneyRedHat

File: image.go [hunk 1]
```diff
@@ -321,10 +321,7 @@ func (i *containerImageRef) NewImageSource(ctx context.Context, sc *types.System
 		}
 		// If we're not re-exporting the data, and we're reusing layers individually, reuse
 		// the blobsum and diff IDs.
-		if !i.exporting && !i.squash && layerID != i.layerID {
-			if layer.UncompressedDigest == "" {
-				return nil, errors.Errorf("unable to look up size of layer %q", layerID)
-			}
+		if !i.exporting && !i.squash && layerID != i.layerID && layer.UncompressedDigest != "" {
```
Comment (TomSweeneyRedHat) [inline image.go:324]: I'm going back and forth.  I'm just wondering if we should keep the if statement still, but replace the return with a debug.  I'm leaning towards no, but not fully  @nalind thoughts?

Comment (nalind) [reply image.go:324]: Well right now, we don't set this information on the mapped layer in the storage library, and this forces us to reconstruct the layer in such a way that its diffID might be different than that of the layer that it was based off of.  We either do that, or we try to redirect ourselves to the corresponding not-mapped layer, and I don't think the storage library gives us quite enough information to do that.

**Comment by nalind** [comment]
LGTM

**Comment by rhatdan** [comment]
Thanks @hshiina

**Comment by TomSweeneyRedHat** [comment]
Just noting that this also fixes a BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1758999

**Pull request merged**